core/storage/wal.rs

```diff
@@ -194,63 +225,176 @@ impl Wal for WalFile {
         pager: &Pager,
         write_counter: Rc<RefCell<usize>>,
     ) -> Result<CheckpointStatus> {
-        let mut shared = self.shared.write().unwrap();
-        for (page_id, _frames) in shared.frame_cache.iter() {
-            // move page from WAL to database file
-            // TODO(Pere): use splice syscall in linux to do zero-copy file page movements to improve perf
-            let page_id = *page_id as usize;
-            if self.ongoing_checkpoint.contains(&page_id) {
-                continue;
-            }
-
-            let page = pager.read_page(page_id)?;
-            if page.is_locked() {
-                return Ok(CheckpointStatus::IO);
+        'checkpoint_loop: loop {
+            let state = self.ongoing_checkpoint.state;
+            log::debug!("checkpoint(state={:?})", state);
+            match state {
+                CheckpointState::Start => {
+                    // TODO(pere): check what frames are safe to checkpoint between many readers!
+                    self.ongoing_checkpoint.min_frame = self.min_frame;
+                    self.ongoing_checkpoint.max_frame = self.max_frame;
+                    self.ongoing_checkpoint.current_page = 0;
+                    self.ongoing_checkpoint.state = CheckpointState::ReadFrame;
+                }
+                CheckpointState::ReadFrame => {
+                    let shared = self.shared.read().unwrap();
+                    if self.ongoing_checkpoint.current_page as usize >= shared.pages_in_frames.len()
```

Comment:
is it a bug if current_page > pages_in_frames.len()? i.e. is it always either less or equal

Reply:
Yes, I guess. It should never happen.

Reply:
maybe add assert and make == the condition?

Reply:
okay